proposals/unions.md

```diff
@@ -0,0 +1,322 @@
+
+``` c#
+if (GetPet() is Dog dog) { ... } // 'Dog dog' is applied to the contents of the Pet value
+if (GetPet() is null) { ... } // 'null' is applied to the contents of the Pet value
+```
+
+If members prescribed by the non-boxing access pattern can be used to implement the pattern behavior, they will. The rules are as follows:
+
+* For a pattern that implies checking for a specific type `T`, if a `TryGetValue(S value)` method is available, and there is an implicit conversion from `T` to `S`, then that method is used to obtain the value. The pattern is then applied to that value. If there is more than one such method, then any where the conversion from `T` to `S` is not a boxing conversion is preferred if available. If there is still more than one method, one is chosen in an implementation-defined manner.
+* Otherwise, for a pattern that implies checking for `null`, if a `HasValue` property is available, that property is used to check if the union value is null.
+* Otherwise, the pattern is applied to the result of accessing the `IUnion.Value` property on the incoming union.
+
+#### Union exhaustiveness
+
+A union type is assumed to be "exhausted" by its case types. This means that a `switch` expression is exhaustive if it handles all of a union's case types:
+
+
+``` c#
+var name = pet switch
+{
+    Dog dog => ...,
+    Cat cat => ...,
+    // No warning about non-exhaustive switch
+};
+```
+
+#### Nullability
+
+The null state of a union's `Value` property is tracked like any other property, with these modifications: 
+
+- For union types where none of the case types are nullable, the default state for `Value` is "not null" rather than "maybe null". 
+- When a union constructor is called (explicitly, through a union conversion or via an implementation of `IUnion<TUnion>.TryCreate(...)`), the new union's `Value` gets the null state of the incoming value.
+- When the non-boxing access pattern's `HasValue` or `TryGetValue(...)` are used to query the contents of a union type (explicitly or via pattern matching), it impacts `Value`'s nullability state in the same way as if `Value` had been checked directly: The null state of `Value` becomes "not null" on the `true` branch.
+
+Even when a union switch is otherwise exhaustive, if the null state of the incoming union's `Value` property is "maybe null", a warning will be given on unhandled null.
+
+``` c#
+Pet pet = GetNullableDog(); // 'pet.Value' is "maybe null"
+var value = pet switch
+{
+    Dog dog => ...,
+    Cat cat => ...,
+    // Warning: 'null' not handled
+}
+```
+
+### Union declarations
+
+Union declarations are a succinct and opinionated way of declaring union types in C#. They declare a struct which uses a single object reference for storage, which means that:
+
+* Any value types among their case types will be boxed on entry.
+* Union values are pointer-length in size, and thus protected from "tearing".
```

Comment:
i'm somewhat wary that this is boxing us into a form we may regret later.

Reply:
(pun intended)

Reply:
Yeah, this is the compromise we reached: The union declaration syntax favors what we believe to be the common case: all the case types are either reference types, or it doesn't really matter that they are boxed. For when that's not true you are in a high-performance scenario and can write your own.

It's quite possible that we'll want to add non-boxing union declarations to the language at some point. Perhaps we'll do it with a `struct` modifier. Our future selves will then have to apply the learnings of the intervening years to pick the content layout strategy that works best for the common case of the performance subset that boxing didn't work for.